sonic-xcvrd/xcvrd/sff_mgr.py

```diff
@@ -0,0 +1,407 @@
+        in a deterministc way, meaning TX is enabled only after host_tx_ready
+        becomes True, and TX will be disabled when host_tx_ready becomes False.
+        This will help eliminate link stability issue and potential interface
+        flap, also turning off TX reduces the power consumption and avoid any
+        lab hazard for admin shut interface.
+
+        Platform can decide whether to enable sff_mgr via platform
+        enable_sff_mgr flag. If enable_sff_mgr is False or not present, sff_mgr
+        will not run. By default, it's disabled.
+
+        There is a pre-requisite for the platforms that
+        enable this sff_mgr feature: platform needs to keep TX in disabled state
+        after module coming out-of-reset, in either module insertion or bootup
+        cases. This is to make sure the module is not transmitting with TX
+        enabled before host_tx_ready is True. No impact for the platforms in
+        current deployment (since they don't enable it explictly.)
+
+        '''
+
+        # CONFIG updates, and STATE_DB for insertion/removal, and host_tx_ready change
+        sel, asic_context = port_mapping.subscribe_port_update_event(self.namespaces, self,
+                                                                     self.PORT_TBL_MAP)
+
+        # This thread doesn't need to expilictly wait on PortInitDone and
+        # PortConfigDone events, as xcvrd main thread waits on them before
+        # spawrning this thread.
+        while not self.task_stopping_event.is_set():
+            # Internally, handle_port_update_event will block for up to
+            # SELECT_TIMEOUT_MSECS until a message is received(in select
+            # function). A message is received when there is a Redis SET/DEL
+            # operation in the DB tables. Upon process restart, messages will be
+            # replayed for all fields, no need to explictly query the DB tables
+            # here.
+            if not port_mapping.handle_port_update_event(
+                    sel, asic_context, self.task_stopping_event, self, self.on_port_update_event):
+                # In the case of no real update, go back to the beginning of the loop
+                continue
+
+            for lport in list(self.port_dict.keys()):
+                if self.task_stopping_event.is_set():
+                    break
+                data = self.port_dict[lport]
+                pport = int(data.get('index', '-1'))
+                channel = int(data.get('channel', '0'))
+                xcvr_type = data.get('type', None)
+                xcvr_inserted = False
+                host_tx_ready_changed = False
+                admin_status_changed = False
+                if pport < 0 or channel < 0:
+                    continue
+
+                if xcvr_type is None:
+                    # TRANSCEIVER_INFO table's 'type' is not ready, meaning xcvr is not present
+                    continue
+
+                # Procced only for 100G/40G
```

Comment:
May be better to remove the reference to speeds since with these FF's also additional speeds can be used - 25/50/200 ...
The code below checks the form factor only